File: android_framework/base/src/main/java/github/tornaco/android/thanos/core/profile/RuleInfo.java
package github.tornaco.android.thanos.core.profile;

import android.os.Parcel;
import android.os.Parcelable;

public final class RuleInfo implements Parcelable {
    private String name;
    private String description;
    private String ruleString;
    private String author;
    private String updateTimeMills;
    private boolean enabled;
    private int format;

    private RuleInfo(Parcel in) {
        name = in.readString();
        description = in.readString();
        ruleString = in.readString();
        author = in.readString();
        updateTimeMills = in.readString();
        enabled = in.readByte() != 0;
        format = in.readInt();
    }

    public RuleInfo(String name,
                    String description,
                    String ruleString,
                    String author,
                    String updateTimeMills,
                    boolean enabled,
                    int format) {
        this.name = name;
        this.description = description;
        this.ruleString = ruleString;
        this.author = author;
        this.updateTimeMills = updateTimeMills;
        this.enabled = enabled;
        this.format = format;
    }

    public static final Creator<RuleInfo> CREATOR = new Creator<RuleInfo>() {
        @Override
        public RuleInfo createFromParcel(Parcel in) {
            return new RuleInfo(in);
        }

        @Override
        public RuleInfo[] newArray(int size) {
            return new RuleInfo[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(name);
        parcel.writeString(description);
        parcel.writeString(ruleString);
        parcel.writeString(author);
        parcel.writeString(updateTimeMills);
        parcel.writeByte((byte) (enabled ? 1 : 0));
        parcel.writeInt(format);
    }
}

File: android_framework/base/src/main/java/github/tornaco/android/thanos/core/profile/RuleInfo.kt
package github.tornaco.android.thanos.core.profile

import android.os.Parcel
import android.os.Parcelable

data class RuleInfo(
    var name: String,
    var description: String,
    var ruleString: String,
    var author: String,
    var updateTimeMills: Long,
    var enabled: Boolean,
    var format: Int
) : Parcelable {

    private constructor(parcel: Parcel) : this(
        parcel.readString(),
        parcel.readString(),
        parcel.readString(),
        parcel.readString(),
        parcel.readLong(),
        parcel.readByte() != 0.toByte(),
        parcel.readInt()
    )

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(name)
        parcel.writeString(description)
        parcel.writeString(ruleString)
        parcel.writeString(author)
        parcel.writeLong(updateTimeMills)
        parcel.writeByte(if (enabled) 1 else 0)
        parcel.writeInt(format)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<RuleInfo> {
        override fun createFromParcel(parcel: Parcel): RuleInfo {
            return RuleInfo(parcel)
        }

        override fun newArray(size: Int): Array<RuleInfo?> {
            return arrayOfNulls(size)
        }
    }

}
